Copy recordPluginMetrics in CycleState.Clone

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

We currently don't copy `recordPluginMetrics` field in CycleState.Clone() so `recordPluginMetrics` in the copied CycleCtate is always false. 
But if we do so, for example, the following part will always be `false`. Is this the expected behavior..?
https://github.com/kubernetes/kubernetes/blob/1007f4974d299991c5cb840dd3ce8f1cc58fbdce/pkg/scheduler/framework/runtime/framework.go#L632

Ref: https://github.com/kubernetes/kubernetes/pull/108724#discussion_r827738293

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix a bug where metrics are not recorded during Preemption(PostFilter).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

denkensk reviewed (commented):
Can you also add an UT or add one case in exist UT for it?

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho commented:
/retest

Huang-Wei commented:
Personally, I'm conservative on using sync.Map, given (1) we seem not to be knowledgeable enough to decide in which situation to use it versus not. Also, the whole k8s codebase doesn't adopt it that much. (2) I'm not sure I fully understand the comments of https://pkg.go.dev/sync#Map. But in any case, CycleState is a generic interface, if using sync.Map can only benefit in the "only-write-the-key-once" case, we'd better _not_ make this assumption as the out-of-tree plugins can use it in any way. A standard map with RWMutex is a neutral option.

sanposhiho commented:
@Huang-Wei your comment is for #108724, right?
Copied your comment in https://github.com/kubernetes/kubernetes/pull/108724#issuecomment-1073192734

alculquicondor commented:
We only Clone the state when we are doing preemption. I think we don't care about metrics within each sub-step of preemption, as preemption itself has an overall latency metric.

Huang-Wei commented:
> @Huang-Wei your comment is for #108724, right?

ah, yes...

Huang-Wei commented:
Hmm.. seems we had a bug that the following metrics were never recorded:

- scheduler_plugin_execution_duration_seconds{extension_point=PreFilterExtensionAddPod}
- scheduler_plugin_execution_duration_seconds{extension_point=PreFilterExtensionRemovePod}

as the **cloned** state always return false on `ShouldRecordPluginMetrics()`.

> I think we don't care about metrics within each sub-step of preemption, as preemption itself has an overall latency metric.

The impact is on RunPreFilterExtensionAddPod() and RunPreFilterExtensionRemovePod(). I think we should copy the value. See the explanation above.

alculquicondor commented:
Right, but should we record the metrics for Filter/Score that run inside preemption?

Huang-Wei commented:
> Right, but should we record the metrics for Filter/Score that run inside preemption?

I think so. Actually `plugin_execution_duration_seconds` is not that useful to diagnose how a pod's overall scheduling looks like. I interpreted it simply to represent a plugin's running at a particular extension point, so it's more "stateless", and thus whether preemption is involved doesn't quite matter.

Huang-Wei commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -41,6 +41,7 @@ func TestCycleStateClone(t *testing.T) {
 		data: data1,
 	}
 	state.Write(key, originalValue)
+	state.SetRecordPluginMetrics(true)
```
We shouldn't set it, this is something we want to test against.

Huang-Wei commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -63,6 +64,10 @@ func TestCycleStateClone(t *testing.T) {
 	}
+
+	if stateCopy.ShouldRecordPluginMetrics() != state.ShouldRecordPluginMetrics() {
+		t.Errorf("cloned copy should have the same recordPluginMetrics value, original state's recordPluginMetrics: %v, copy state's recordPluginMetrics %v", state.ShouldRecordPluginMetrics(), stateCopy.ShouldRecordPluginMetrics())
```
```suggestion
		t.Errorf("unexpected recordPluginMetrics: want %v, but got %v", state.ShouldRecordPluginMetrics(), stateCopy.ShouldRecordPluginMetrics())
```

Huang-Wei commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -41,6 +41,7 @@ func TestCycleStateClone(t *testing.T) {
 		data: data1,
 	}
 	state.Write(key, originalValue)
+	state.SetRecordPluginMetrics(true)
```
can we test both true and false?

sanposhiho replied:
👍 
Changed the whole test to be table-driven test and added the test for that as a test case.

sanposhiho commented:
/retest

sanposhiho commented:
/retest

Huang-Wei commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +31,88 @@ func (f *fakeData) Clone() StateData {
+// createCycleStateWithFakeData creates *CycleState with fakeData.
+// The given data is used in stored fakeData.
+func createCycleStateWithFakeData(data string, recordPluginMetrics bool) *CycleState {
+	c := NewCycleState()
```
missing set `recordPluginMetrics`:

```go
c.SetRecordPluginMetrics(recordPluginMetrics)
```

Huang-Wei commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +31,88 @@ func (f *fakeData) Clone() StateData {
+		t.Run(tt.name, func(t *testing.T) {
+			state := tt.state
+			stateCopy := state.Clone()
+			if tt.wantClonedState == nil {
```
I feel the logic can be simplified by leveraging `cmp.Diff()`. Does https://gist.github.com/Huang-Wei/5f4bd967d3adf3d738531c09214bc307 look better?

sanposhiho replied:
Looks nice. Thanks

sanposhiho committed:
Copy recordPluginMetrics in CycleState.Clone

sanposhiho commented:
@Huang-Wei 
Could you please take a look at this again 🙏

sanposhiho commented:
/retest

sanposhiho commented:
Rebased to resolve the conflict.

@Huang-Wei 
I have reverted the test back to what it was before I applied [your recommendations](https://github.com/kubernetes/kubernetes/pull/108727#discussion_r836669667) since it's difficult to use `cmp.Diff` to compare `sync.Map`. (If there is a better way to use `cmp.Diff`, please let me know. 🙏  )

sanposhiho commented:
/retest

sanposhiho commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +31,89 @@ func (f *fakeData) Clone() StateData {
+		t.Run(tt.name, func(t *testing.T) {
+			state := tt.state
+			stateCopy := state.Clone()
+			if tt.wantClonedState == nil {
```
changed to use cmp.Diff (with cmp.Comparer to compare `sync.Map`.

alculquicondor commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +34,87 @@ func (f *fakeData) Clone() StateData {
+				}
+
+				// convert type to compare v1 and v2
+				typedV1, ok1 := v1.(*fakeData)
```
can you call `cmp.Equal(v1, v2)` here?

alculquicondor commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +34,79 @@ func (f *fakeData) Clone() StateData {
+				}
+				return true
+			})
+			return isEqual
```
also need to check if the number of elements is the same

sanposhiho replied:
Definitely. Fixed.

alculquicondor commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +34,92 @@ func (f *fakeData) Clone() StateData {
 
-	if v, ok := valueCopy.(*fakeData); ok && v.data != data1 {
-		t.Errorf("cloned copy should not change, got %q, expected %q", v.data, data1)
+			if countA != countB {
```
return isEqual && countA == countB

alculquicondor replied:
this?

sanposhiho replied:
🙇  fixed.

alculquicondor commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +34,92 @@ func (f *fakeData) Clone() StateData {
-		t.Errorf("clone expected to be nil")
+			stateCopy.Write(key, &fakeData{data: "modified"})
+			if diff := cmp.Diff(state, stateCopy, cmpOpts...); diff == "" {
+				t.Errorf("Unexpected cloned state (-want, +got):\n%s", diff)
```
This message is weird. We want an error if they are equal, in which case `diff == ""`. Maybe just use `cmp.Equal` in this case?

sanposhiho replied:
Fixed the error message and change to use `cmp.Equal`

sanposhiho commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +34,98 @@ func (f *fakeData) Clone() StateData {
-	}
+	cmpOpts := []cmp.Option{
+		cmp.AllowUnexported(CycleState{}),
+		cmp.Comparer(func(a, b *sync.Map) bool {
```
changed `Comparer` from `sync.Map` to `*sync.Map`. 
This resolves the following go vet issue.

```
func passes lock by value: sync.Map contains sync.Mutex
```

Huang-Wei replied:
I'd pursue working the govet error around cuz defining a pointer of sync.Map sounds weird...

alculquicondor replied:
I don't know how we can avoid the error. The comparer would receive elements by copy, and it's unsafe to copy a lock.

sanposhiho replied:
Hmm, I don't have an idea to avoid the issue from go vet (other than using `*sync.Map`).

One way is to give up using `cmp.Diff` and `cmp.Equal`... 
In that case, the test would be longer, but would look like this:
https://github.com/kubernetes/kubernetes/commit/55c5d51286ff947e3c2233d92b26aa7911f6222e

@alculquicondor @Huang-Wei 
Which do you prefer, the current implementation of this branch or the implementation without `cmp.Diff` like https://github.com/kubernetes/kubernetes/commit/55c5d51286ff947e3c2233d92b26aa7911f6222e?

alculquicondor replied:
My preference is for the implementation in this PR, but I won't oppose to the other if Wei feels strongly about the pointer change.

Huang-Wei replied:
@sanposhiho Given the `go vet` is trying to avoid copy values containing mutex. We can tweak the test like this: https://gist.github.com/Huang-Wei/5f4bd967d3adf3d738531c09214bc307#file-108727-diff.

Huang-Wei replied:
Yup, we can pursue debug readability over being restricted with cmp.Diff.

sanposhiho replied:
👍 
Change not to use cmp.Diff in a959c3bf7c96911361f4688895494dcf7855b9dc.

sanposhiho commented:
/retest

alculquicondor commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +34,98 @@ func (f *fakeData) Clone() StateData {
+
+			// we need to compare *sync.Map,
+			// because we cannot define cmp.Comparer for sync.Map or CycleState due to the issue from go vet.
+			if diff := cmp.Diff(&state.storage, &stateCopy.storage, cmpOpts...); diff != "" {
```
Now I think nothing is checking `recordPluginMetrics`?

is it a lot of changes if you make CycleState.storage a pointer?

sanposhiho replied:
Fixed.

alculquicondor reviewed (commented):
/lgtm
/approve

This can wait for 1.25

sanposhiho commented:
/retest

Huang-Wei commented on pkg/scheduler/framework/cycle_state_test.go:
```diff
@@ -31,44 +32,120 @@ func (f *fakeData) Clone() StateData {
+			stateCopy := state.Clone()
+
+			if isEqual, msg := isCycleStateEqual(stateCopy, tt.wantClonedState); !isEqual {
+				t.Errorf("unexpected cloned state: %v\n", msg)
```
no need to append `\n` (similar as L147)

sanposhiho replied:
Fixed.

Huang-Wei commented:
/lgtm

sanposhiho commented:
@Huang-Wei @alculquicondor 

Is it okay to `/unhold`?

alculquicondor reviewed (commented):
/hold cancel

sanposhiho commented:
/retest
